src/Composer/DependencyResolver/PoolBuilder.php

```diff
@@ -419,32 +419,37 @@ private function loadPackagesMarkedForLoading(Request $request, array $repositor
             $this->loadedPackages[$name] = $constraint;
         }
 
-        $packageBatch = $this->packagesToLoad;
+        // Load packages in chunks of 50 to prevent memory usage build-up due to caches of all sorts
+        $packageBatches = array_chunk($this->packagesToLoad, 50, true);
         $this->packagesToLoad = [];
 
         foreach ($repositories as $repoIndex => $repository) {
-            if (0 === \count($packageBatch)) {
-                break;
-            }
+            foreach ($packageBatches as $packageBatch) {
+                if (0 === \count($packageBatch)) {
+                    break;
+                }
```

Comment:
This needs to be moved below the packageBatches foreach probably.. or at least it needs to be revisited, because the point was to abort the repository loop when we're done with all packages.

I guess ideally you'd re-batch all packages that are left into a new set of 50-packages batches after every packageBatches foreach, so that every repo gets queried with 50 packages at once consistently.

```suggestion
                if (0 === \count($packageBatch)) {
                    break;
                }
```

Reply:
I can't wrap my head around a solution for that for now.